Sources/NIO/BaseSocketChannel.swift

```diff
@@ -469,32 +475,82 @@ class BaseSocketChannel<SocketType: BaseSocketProtocol>: SelectableChannel, Chan
         return try self.socket.remoteAddress()
     }
 
+    private final func flushNowAndRegisterIfNecessary(mayHaveReentered: Bool) {
+        assert(self.isOpen)
+        assert(self.lifecycleManager.isActive)
+
+        switch self.flushNowState {
+        case .inFlushNowCallingOut:
+            assert(mayHaveReentered)
+            // We're calling out from `flushNow` after `writeSocket`'s write loop has finished. We therefore need to
+            // replay the `flushNow` later because we otherwise might not write a flushed write.
+            self.eventLoop.execute {
+                self.flushNowAndRegisterIfNecessary(mayHaveReentered: false)
+            }
+            return
+        case .inFlushNowWriting:
+            assert(mayHaveReentered)
+            // Nothing to do here, we're still in `writeSocket`'s write loop so the previously flushed writes will be
+            // picked up.
+            return
+        case .notInFlushNow:
+            () // Okay, let's actually do this.
+        }
+
+        if !self.isWritePending() {
+            switch self.flushNow() {
+            case .register:
+                assert(self.lifecycleManager.isPreRegistered)
+                self.registerForWritable()
+            case .unregister:
+                // Nothing to do because we weren't registered when we called `flushNow`. We also know that `flushNow`
+                // isn't called re-entrantly so we know we still don't have a pending write.
+                assert(!self.isWritePending())
+                ()
+            }
+        }
+    }
+
     /// Flush data to the underlying socket and return if this socket needs to be registered for write notifications.
     ///
-    /// - returns: If this socket should be registered for write notifications. Ie. `IONotificationState.register` if _not_ all data could be written, so notifications are necessary; and `IONotificationState.unregister` if everything was written and we don't need to be notified about writability at the moment.
+    /// - warning: Only call this method i
```

Comment:
there is something missing... I guess "only call this method in the eventloop" ?

Reply:
whoops, thank you!

Reply:
fixed.